packages/twenty-server/src/engine/core-modules/ask-ai/ask-ai.service.ts

```diff
@@ -0,0 +1,181 @@
+    private readonly llmPromptTemplateService: LLMPromptTemplateService,
+    private readonly llmTracingService: LLMTracingService,
+    private readonly objectMetadataService: ObjectMetadataService,
+  ) {}
+
+  private async getRecordMetadataById(
+    workspaceId: string,
+    workspaceDataSource: DataSource,
+    sqlQueryResult: Record<string, any>[],
+  ) {
+    const uuids = sqlQueryResult
+      .flatMap((row) => Object.values(row))
+      .filter((value) => isUUID(value, 4));
+
+    const objectMetadataEntities =
+      await this.objectMetadataService.findManyWithinWorkspace(workspaceId, {
+        where: { isSystem: false, isActive: true },
+      });
+
+    const columnNamesByStandardObjectName = {
+      opportunity: ['id', 'name'],
+      company: ['id', 'name', 'domainName'],
+      person: ['id', 'nameFirstName', 'nameLastName'],
+    };
+
+    const customObjectColumnNames = ['id', 'name'];
+
+    let recordMetadataById: RecordMetadataById = {};
+
+    for (const { nameSingular, isCustom } of objectMetadataEntities) {
+      const tableName = computeTableName(nameSingular, isCustom);
+
+      const columnNames = isCustom
+        ? customObjectColumnNames
+        : columnNamesByStandardObjectName[nameSingular];
+
+      const columnNamesString =
+        columnNames.map((col) => `"${col}"`).join(', ') || '';
+
+      // Security? GraphQL?
+      const recordMetadataSqlQuery = `SELECT ${columnNamesString} FROM "${tableName}" WHERE "id" = ANY($1);`;
+
+      const result: {
+        id: string;
+        name: string;
+        domainName?: string;
+        nameFirstName?: string;
+        nameLastName?: string;
+      }[] = await this.workspaceQueryRunnerService.executeSQL(
+        workspaceDataSource,
+        workspaceId,
+        recordMetadataSqlQuery,
+        [uuids],
+      );
+
+      recordMetadataById = result.reduce((recordMetadataById, row) => {
+        return {
+          ...recordMetadataById,
+          [row.id]: {
+            objectNameSingular: nameSingular,
+            ...omit(row, 'id'),
+          },
+        };
+      }, recordMetadataById);
+    }
+
+    return recordMetadataById;
+  }
+
+  async query(
+    userId: string,
+    userEmail: string,
+    workspaceId: string,
+    text: string,
+  ): Promise<AskAIQueryResult> {
+    const workspaceSchemaName =
+      this.workspaceDataSourceService.getSchemaName(workspaceId);
+
+    const workspaceDataSource =
+      await this.workspaceDataSourceService.connectToWorkspaceDataSource(
+        workspaceId,
+      );
+
+    workspaceDataSource.setOptions({ schema: workspaceSchemaName });
+
+    const db = await SqlDatabase.fromDataSourceParams({
+      appDataSource: workspaceDataSource,
+      sampleRowsInTableInfo: 0,
+    });
+
+    const promptTemplate =
+      await this.llmPromptTemplateService.getPromptTemplate(
+        LLMPromptTemplateEnvVar.AskAI,
+      );
+
+    const removeSQLMarkdown: RunnableFunc<string, string> = (input) =>
```

Comment:
We talked about "forceJson" - I'm surprised there isn't a clear way to do this, this seems very incertain as I see nothing matching in the code that would indicate the LLM is going to output it in this format? I assume you introduce something in each prompt? 
 Could you separate the formatting instruction from the rest of the prompt at least to make sure they are coherent with your parsing below? (e.g. if you decide to give formatting instruction JSON, then parse JSON in the code)

Reply:
Using JSON mode now!